File: 987-vertical-order-traversal-of-a-binary-tree/987-vertical-order-traversal-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        nodes = defaultdict(lambda: defaultdict(list))
        
        def pre_order(node: TreeNode | None, row: int, col: int):
            if node.left:
                pre_order(node.left, row + 1, col - 1)
            
            nodes[col][row].append(node.val)
            
            if node.right:
                pre_order(node.right, row + 1, col + 1)
        
        pre_order(root, 0, 0)

        res = []
        for col in sorted(nodes):
            lst = []
            for row in sorted(nodes[col]):
                lst.extend(sorted(nodes[col][row]))
            res.append(lst)
            
        return res

File: 987-vertical-order-traversal-of-a-binary-tree/987-vertical-order-traversal-of-a-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var verticalTraversal = function(root) {
  // nodes := { 
  //   col1: { 
  //     row1: [1, 2,...], 
  //     row2: [..] 
  //   }, 
  //   col2: {..} 
  // }
  const nodes = {}
  
  ;(function preOrder(node, row, col) {
    if (node.left) {
      preOrder(node.left, row + 1, col - 1)
    }
    
    if (!nodes.hasOwnProperty(col)) {
      nodes[col] = { [row]: [node.val] }
    } else if (!nodes[col].hasOwnProperty(row)) {
      nodes[col][row] = [node.val]
    } else {
      nodes[col][row].push(node.val);
    }
    
    if (node.right) {
      preOrder(node.right, row + 1, col + 1)
    }
  })(root, 0, 0)
  
  const ascending = (a, b) => a - b
  
  const res = []
  for (const col of Object.keys(nodes).sort(ascending)) {
    const vert = []
    for (const row of Object.keys(nodes[col]).sort(ascending)) {
      vert.push.apply(vert, nodes[col][row].sort(ascending))
    }
    res.push(vert)
  }
  
  return res
};
